xrpl4j-core/src/main/java/org/xrpl/xrpl4j/crypto/keys/PrivateKey.java

```diff
@@ -32,41 +33,189 @@
   public static PrivateKey of(final UnsignedByteArray value) {
-    return new PrivateKey(value);
+    // Assumption: Any developer using this method before it was deprecated (i.e., v3.2.1 and before) will likely have
+    // expected `value` to have been prefixed. However, until this method was "fixed" in v3.2.2, a developer might have
+    // used this method incorrectly because (1) this method had no length check and (2) the computation of the KeyType
+    // naively inspected the first byte, and defaulted to secp256k1 if the first byte was `0xED`. For example, in that
+    // world, padding in a one-byte value of `0x00` would have produced a `PrivateKey` object with invalid byte data.
+    return fromPrefixedBytes(value);
   }
 
   /**
    * Required-args Constructor.
    *
    * @param value An {@link UnsignedByteArray} for this key's value.
    */
-  private PrivateKey(final UnsignedByteArray value) {
-    this.value = Objects.requireNonNull(value);
+  private PrivateKey(final UnsignedByteArray value, final KeyType keyType) {
+    Objects.requireNonNull(value); // <-- Check not-null first.
+    this.keyType = Objects.requireNonNull(keyType);
+
+    // We assert this precondition here because this is a private constructor that can be fully tested via unit test,
+    // so this precondition should never be violated, and if it is, then it's a bug in xrpl4j.
+    Preconditions.checkArgument(
+      value.length() == 32,
+      "Byte values passed to this constructor must be 32 bytes long, with no prefix."
+    );
+
+    this.value = UnsignedByteArray.of(value.toByteArray()); // <- Always copy to ensure immutability
   }
 
   /**
    * Accessor for the key value, in binary (Note: will be 33 bytes).
    *
    * @return An instance of {@link UnsignedByteArray}.
+   *
+   * @deprecated Prefer {@link #valueWithPrefixedBytes()} or {@link #valueWithNaturalBytes()} instead.
    */
+  @Deprecated
   public UnsignedByteArray value() {
-    return UnsignedByteArray.of(value.toByteArray());
+    // Check for empty value, which can occur if this PrivateKey is "destroyed" but still in memory.
+    if (value.length() == 0) {
+      return UnsignedByteArray.empty();
+    } else {
+      // This is technically wrong (because `value()` had an ambiguous meaning prior to fixing #486), but this mirrors
+      // what's in v3 prior to fixing #486, and will be fixed in v4 once the deprecated `.value()` is removed.
+      return this.valueWithPrefixedBytes();
+    }
+  }
+
+  /**
+   * Accessor for the byte value in {@link #value()} but in a more natural form (i.e., the size of the returned value
+   * will be 32 bytes). Natural ed25519 or secp256k1 private keys will ordinarily contain only 32 bytes. However, in
+   * XRPL, private keys are represented with a single-byte prefix (i.e., `0xED` for ed25519 and `0x00` for secp256k1
+   * keys).
+   *
+   * @return An instance of {@link UnsignedByteArray}.
+   */
+  public UnsignedByteArray valueWithNaturalBytes() {
```

Comment:
The name of this method seems off for some reason... what if we just call it `naturalBytes()` or `naturalBytesValue()`?

Reply:
I actually like `naturalBytes`, but if we do that then I suggest we rename the other one to `prefixedBytes`. 

This seems actually preferable overall. One minor nit is developer ergonomics: it won't be possible to type `key.value...` and have auto-complete work.

But that's such a small thing, and perhaps it's better to move away from the `value` nomenclature. WDYT?

Reply:
Agree, I like `naturalBytes` and `prefixedBytes`

Reply:
Fixed in 9cf65c4a869426120fa30e7dbf750b6ac659fa75